### PR TITLE
Transformers

### PR DESCRIPTION
**Pull request by saidbleik**
### Checklist:

- [ ] My code follows the code style of this project, as detailed in our [contribution guidelines](../CONTRIBUTING.md).
- [ ] I have added tests.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: examples/text_classification/tc_mnli.ipynb [hunk 1]
```diff
@@ -0,0 +1,6283 @@
+{
```
Comment (awaemmanuel) [inline examples/text_classification/tc_mnli.ipynb]: @saidbleik we should rename this to `tc_transformers_mnli.ipynb`

File: utils_nlp/models/transformers/common.py [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+                outputs = self.model(**inputs)
+                logits = outputs[0]
+
+            yield logits.detach().cpu().numpy()
+
+    def save_model(self):
+        output_model_dir = os.path.join(self.cache_dir, "fine_tuned")
+
+        if not os.path.exists(self.cache_dir):
+            os.makedirs(self.cache_dir)
```
Comment (miguelgfierro) [inline utils_nlp/models/transformers/common.py]: minor thing, this is equivalent to:
```
os.makedirs(self.cache_dir, exist_ok=True)
```

**Comment by miguelgfierro** [comment]
hey guys, this looks really nice. I did a first pass and will continue tomorrow

File: tools/generate_conda_file.py [hunk 3]
```diff
@@ -64,6 +65,7 @@
     "azureml-mlflow": "azureml-mlflow>=1.0.43.1",
     "black": "black>=18.6b4",
     "cached-property": "cached-property==1.5.1",
+    "jsonlines": "jsonlines>=1.2.0",
     "papermill": "papermill>=1.0.1",
     "nteract-scrapbook": "nteract-scrapbook>=0.2.1",
     "pydocumentdb": "pydocumentdb>=2.3.3",
```
[the remark this answers is not in the document]
Comment (saidbleik) [reply tools/generate_conda_file.py]: Only for now. We can remove those once the notebook examples use the new utils.

File: examples/text_classification/tc_mnli_transformers.ipynb [hunk 4]
```diff
@@ -0,0 +1,6283 @@
+{
```
Comment (miguelgfierro) [inline examples/text_classification/tc_mnli_transformers.ipynb]: <p>like in the previous notebook, we can consider to use here <code>TemporaryDirectory()</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/nlp/pull/434/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (miguelgfierro) [inline examples/text_classification/tc_mnli_transformers.ipynb]: <p>would you mind to move this list to the parameters cell? that way we can control how many models we want to use in the tests</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/nlp/pull/434/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (miguelgfierro) [inline examples/text_classification/tc_mnli_transformers.ipynb]: <p>same comment as before, there are too many logs, the notebook would look nicer without them</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/nlp/pull/434/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

File: utils_nlp/dataset/pytorch.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+import collections
+from torch.utils.data import Dataset
+
+
+class SCDataSet(Dataset):
+    def __init__(self, df, text_col, label_col):
```
Comment (miguelgfierro) [inline utils_nlp/dataset/pytorch.py]: in the init, we can also add docstrings and they are parsed by sphinx thanks to a command we added https://github.com/microsoft/nlp/blob/master/docs/source/conf.py#L229

File: utils_nlp/eval/evaluate_question_answering.py [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+        common = collections.Counter(gold_toks) & collections.Counter(pred_toks)
+        num_same = sum(common.values())
+        if len(gold_toks) == 0 or len(pred_toks) == 0:
+            # If either is no-answer, then F1 is 1 if they agree, 0 otherwise
+            return int(gold_toks == pred_toks)
+        if num_same == 0:
+            return 0
+        precision = 1.0 * num_same / len(pred_toks)
+        recall = 1.0 * num_same / len(gold_toks)
+        f1 = (2 * precision * recall) / (precision + recall)
```
[the remark this answers is not in the document]
Comment (saidbleik) [reply utils_nlp/eval/evaluate_question_answering.py]: This is slightly different. Here, a match counts irrespective of order of tokens. sklearn's F1 expects a 1 to 1 mapping between actual and predicted.

File: utils_nlp/models/transformers/common.py [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+            logger.info("Loading cached model from {}".format(load_model_from_dir))
+            self.model = model_class[model_name].from_pretrained(
+                load_model_from_dir, num_labels=num_labels
+            )
+
+    @property
+    def model_name(self):
+        return self._model_name
+
+    @model_name.setter
```
[the remark this answers is not in the document]
Comment (hlums) [reply utils_nlp/models/transformers/common.py]: All the time... Because the model name determines which pre-trained model is loaded.

[the remark this answers is not in the document]
Comment (saidbleik) [reply utils_nlp/models/transformers/common.py]: right, I think we shouldn't as the constructor loads a pretrained model based on the initial model_name. @hlums

File: utils_nlp/models/transformers/common.py [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+                "weight_decay": weight_decay,
+            },
+            {
+                "params": [
+                    p for n, p in self.model.named_parameters() if any(nd in n for nd in no_decay)
+                ],
+                "weight_decay": 0.0,
+            },
+        ]
+        optimizer = AdamW(optimizer_grouped_parameters, lr=learning_rate, eps=adam_epsilon)
```
[the remark this answers is not in the document]
Comment (saidbleik) [reply utils_nlp/models/transformers/common.py]: Only problem is that these are part of the transformers package. Not sure if user-defined PyTorch optimizers and schedulers would work.

File: utils_nlp/models/transformers/common.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (saidbleik) [reply utils_nlp/models/transformers/common.py]: but the user shouldn't be able to set it again after that. they shouldn't be able to set it to something different from the loaded one.

[the remark this answers is not in the document]
Comment (hlums) [reply utils_nlp/models/transformers/common.py]: @miguelgfierro Sorry I didn't get your question at first. Said and I discussed and he will make it a read-only property.

File: utils_nlp/models/transformers/common.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (miguelgfierro) [reply utils_nlp/models/transformers/common.py]: `AdamW` is just a subclass of `pytorch.Optimizer` https://github.com/huggingface/transformers/blob/master/transformers/optimization.py samething with the lr scheduler. The user would be able to add any optimizer or scheduler. I can make the change or you can do it if you prefer

File: utils_nlp/models/transformers/common.py [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+                if gradient_accumulation_steps > 1:
+                    loss = loss / gradient_accumulation_steps
+
+                if step % 10 == 0 and verbose:
+                    tqdm.write("Loss:{:.6f}".format(loss / train_batch_size))
+
+                if fp16:
+                    with amp.scale_loss(loss, optimizer) as scaled_loss:
+                        scaled_loss.backward()
+                    torch.nn.utils.clip_grad_norm_(amp.master_params(optimizer), max_grad_norm)
```
[the remark this answers is not in the document]
Comment (miguelgfierro) [reply utils_nlp/models/transformers/common.py]: I'm looking at this. I saw different options for this:
- [0.5](https://github.com/yunjey/pytorch-tutorial/blob/master/tutorials/02-intermediate/language_model/main.py#L81)
- [0.25](https://github.com/pytorch/examples/blob/f490f5436958d4fea0250e478b01d0711a09a41a/word_language_model/main.py#L24)
- [1](https://stackoverflow.com/a/54716953)
- [20](https://zhuanlan.zhihu.com/p/76459295)
- [fairseq considers 0](https://github.com/facebookresearch/pytext/blob/master/pytext/optimizer/fp16_optimizer.py#L295) but if we look at the implementation I think there is something weird because having max_norm 0 would make all the [parameters 0](https://pytorch.org/docs/stable/_modules/torch/nn/utils/clip_grad.html)

File: utils_nlp/models/transformers/common.py [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+        if seed is not None:
+            Transformer.set_seed(seed, n_gpu > 0)
+
+        train_batch_size = per_gpu_train_batch_size * max(1, n_gpu)
+        train_sampler = (
+            RandomSampler(train_dataset) if local_rank == -1 else DistributedSampler(train_dataset)
+        )
+        train_dataloader = DataLoader(
+            train_dataset, sampler=train_sampler, batch_size=train_batch_size
+        )
```
[the remark this answers is not in the document]
Comment (saidbleik) [reply utils_nlp/models/transformers/common.py]: Agree. Initially I thought this would save us from computing the batch size during the preprocessing step as we only need to pass it to fit(). I'll change this.

File: utils_nlp/models/transformers/common.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (saidbleik) [reply utils_nlp/models/transformers/common.py]: I'm guessing the different values here have minimal impact on the performance, as long as they are within an acceptable range. I don't think we can be sure without some hyperparameter tuning.

[the remark this answers is not in the document]
Comment (miguelgfierro) [reply utils_nlp/models/transformers/common.py]: yeah, makes sense, also we would need to see the profile with tensorboard. We can leave it as 1 and hope that it doesn't mess up the training
